### ISSUE TITLE
Losing focus on enter magic code section

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#23925]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. go to new.expensify.com
2. enter email and hit enter
3. go to another window of chrome to see the magic code from email
4. come back and observe the first input for magic code is highlighted(apparently, is having focus)
5. start typing the magic code
6. not able to input code (result 1)
7. click anywhere on the page to get back focus in this window
8. start typing the magic code
9. still not able to input code (result 2)

## Expected Result: 
Either the input field should not be highlighted (as it is actually not having the focus) or it should actually get focus and whatever code user types should display.
## Actual Result: 
Input field is highlighted but whatever code user types is not displayed in the input field.

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:**
**Reproducible in staging?:** Needs reproduction (Magic code can be typed and not able to reproduce the issue)
**Reproducible in production?:** Needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/3d9dd694-8053-499b-8c3f-06a0e4a667a7

**Expensify/Expensify Issue URL:**
**Issue reported by:** @himanshuragi456
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690476556105309

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Pujan92** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Focus is not losing in magic code when we shift focus to other content

### What is the root cause of that problem?
We are not resetting the focusedIndex on the blur of the magic code input.

### What changes do you think we should make in order to solve the problem?
We can add `onBlur` event for magic code TextInput where we can call `blurMagicCodeInput` which will reset the `focusedIndex`.
https://github.com/Expensify/App/blob/dc3270005e924e49a890028bd4b3ef094a939227/src/components/MagicCodeInput.js#L313

**Comment by twisterdotcom** [comment]
I still haven't been able to reproduce. Everytime I navigate back to the magic code screen, I can type a number and it adds it to the first input.

Is there something special I have to do to make it lose the focus but still show with a green highlight?

**Comment by twisterdotcom** [comment]
I still don't get it. I try to sign in. I get to the magic code section. I completely switch Chrome windows and focus in the other one. Then, I navigate back to the magic code window and I don't click anywhere inside the window. I type `777` and it enters `777` into the magic code field.

https://github.com/Expensify/App/assets/9133401/3c7c75e5-cede-41a8-8da0-262002c46a42

**Comment by Pujan92** [comment]
> @twisterdotcom you can click on outside/any text of the screen to shift focus.

https://github.com/Expensify/App/assets/14358475/ef1e03ef-12ce-4a9e-8018-94d4f456daad

**Comment by twisterdotcom** [comment]
Ah odd. Okay, yes this is a thing. This is so minor though.

**Comment by joh42** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
There are two problems here:
1. When a user clicks outside of the [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js), the line under the previously focused input remains green, indicating that the user can still type.
2. The [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js) on the [SignInPage](https://github.com/joh42/expensify-app/blob/main/src/pages/signin/SignInPage.js) is not given focus when a user comes back to that browser tab. Most of the time the [MagicCodeInput ](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js)retains focus when a user navigates away, but as we can see in this issue, that is not always the case, leading to inconsistent behavior.
### What is the root cause of that problem?
The root causes are:
1. We do not modify the `focusIndex` when the [TextInput](https://github.com/joh42/expensify-app/tree/main/src/components/TextInput) components are blurred.
2. We do not give the [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js) focus when a user comes back to the [SignInPage](https://github.com/joh42/expensify-app/blob/main/src/pages/signin/SignInPage.js).
### What changes do you think we should make in order to solve the problem?

There have already been proposals that include a solution to the first problem. Specifically, to reset the `focusedIndex` in the [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js) on blur. I propose the same solution to solve the first problem.

To solve the second problem however, I suggest we listen to both visibility changes using the [Visibility lib](https://github.com/joh42/expensify-app/blob/main/src/libs/Visibility/index.js) and window focus events. When the `visibilityState` changes to visible, and/or the window is focused, we focus the [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js). That way we can be sure that the [MagicCodeInput](https://github.com/joh42/expensify-app/blob/main/src/components/MagicCodeInput.js) component is always focused when a user returns to the [SignInPage](https://github.com/joh42/expensify-app/blob/main/src/pages/signin/SignInPage.js), allowing users to promptly enter the magic code they just received. 


### What alternative solutions did you explore? (Optional)

**Comment by twisterdotcom** [comment]
@situchan what do we think of @joh42's proposal?

**Comment by twisterdotcom** [comment]
bump @situchan.

**Comment by situchan** [comment]
> Either the input field should not be highlighted (as it is actually not having the focus) or it should actually get focus and whatever code user types should display.

@twisterdotcom I'd like to confirm the expected result here. Former one or latter one?

**Comment by twisterdotcom** [comment]
>it should actually get focus and whatever code user types should display.

This is ideal obviously, the avoid the user having to do an extra click. If it's quicker and easier to just not highlight it, then let's do that.

Given this is so insanely minor, let's do the easiest least time-consuming fix to make the UX less confusing.

**Comment by situchan** [comment]
Personally I think it's better UX to focus automatically and type in previously focused cell after click any other area which makes it blur and then input any number, similar to composer behavior. Because it's easily got blurred and user needs to move mouse manually to focus on cell.
Btw, I suggest to put this on hold for #18244.

**Comment by twisterdotcom** [comment]
Cool, holding and downgrading to weekly for now then. Thanks @situchan

**Comment by twisterdotcom** [comment]
Apologies but I am finally going on Parental Leave. I am reassigning to another member of the team.

**Comment by situchan** [comment]
still on hold

**Comment by tjferriss** [comment]
I'm updating to weekly to align with issue we're holding on https://github.com/Expensify/App/issues/18244

**Comment by tjferriss** [comment]
It looks like the related PR may get merged this week https://github.com/Expensify/App/pull/24012#issuecomment-1693798895

**Comment by tjferriss** [comment]
We're still waiting on the related PR.

**Comment by tjferriss** [comment]
Still waiting on the related PR

**Comment by tjferriss** [comment]
@situchan it looks like the related PR has been deployed and is now in the regression period. Can we pick this back up?

Also, @twisterdotcom do you want to take this back now that you've returned from parental leave?

**Comment by situchan** [comment]
Unfortunately the holding PR was reverted due to several regressions.
So we should still hold more.

**Comment by twisterdotcom** [comment]
Still held.

**Comment by twisterdotcom** [comment]
Still held.

**Comment by twisterdotcom** [comment]
@situchan we still holding this?

**Comment by situchan** [comment]
We can remove hold.
@himanshuragi456 can you confirm this is still reproducible?

**Comment by twisterdotcom** [comment]
I cannot recreate this now:

https://github.com/Expensify/App/assets/9133401/a8c253c9-5d56-4d12-9ba7-1ad34f42f440
